infra.cpp

```
@@ -0,0 +1,179 @@
+                // Skip block if there is a parsing error
+                blockLines.clear();
+                continue;
+            }
+            blockchain.push_back(b); // Add the block to the vector
+            blockLines.clear();
+        }
+    }
+    file.close();
+}
+
+// Prints all blocks in the blockchain in the required format (no quotes around values).
+// Prints an arrow between blocks for visual separation.
+void print_db() {
+    if (blockchain.empty()) {
+        std::cout << "Blockchain is empty. Please run load_db() first." << std::endl;
+        return;
+    }
+    for (size_t i = 0; i < blockchain.size(); ++i) {
+        const auto& b = blockchain[i];
+        // Print all block fields, one per line
+        std::cout << "hash: " << b.hash << std::endl;
+        std::cout << "height: " << b.height << std::endl;
+        std::cout << "total: " << b.total << std::endl;
+        std::cout << "time: " << b.time << std::endl;
+        std::cout << "relayed_by: " << b.relayed_by << std::endl;
+        std::cout << "prev_block: " << b.prev_block << std::endl;
+        // Print arrow only if not the last block
+        if (i != blockchain.size() - 1) {
+            std::cout << "    |\n    v\n" << std::endl;
+        }
+    }
+}
+
+// Finds and prints a block by its hash value.
+// Assumes blockchain is already loaded into memory.
+void find_block_by_hash(const char* hash) {
+    for (const auto& b : blockchain) {
+        if (b.hash == hash) {
+            std::cout << "hash: " << b.hash << std::endl;
+            std::cout << "height: " << b.height << std::endl;
+            std::cout << "total: " << b.total << std::endl;
+            std::cout << "time: " << b.time << std::endl;
+            std::cout << "relayed_by: " << b.relayed_by << std::endl;
+            std::cout << "prev_block: " << b.prev_block << std::endl;
+            return;
+        }
+    }
+    std::cout << "Block with hash '" << hash << "' not found." << std::endl;
```

Better to throw an exception and print the error from the calling function